bin/runtime/src/migrations/custom_nomination_pools.rs

```diff
@@ -0,0 +1,139 @@
+pub use nomination_pools::CustomMigrateToV2;
+
+mod nomination_pools {
+    use codec::{Decode, Encode, Error, Input};
+    use frame_support::{
+        log,
+        traits::{OnRuntimeUpgrade, StorageVersion},
+    };
+    use pallet_nomination_pools::{
+        BalanceOf, BondedPools, Config, Metadata, Pallet, PoolId, PoolMember, PoolMembers,
+        ReversePoolIdLookup, RewardPool, RewardPools, SubPoolsStorage,
+    };
+    use sp_core::{Get, U256};
+    use sp_std::{collections::btree_set::BTreeSet, prelude::*};
+
+    use crate::sp_api_hidden_includes_construct_runtime::hidden_include::dispatch::GetStorageVersion; // sick
+    use crate::Weight;
+
+    #[derive(Decode)]
+    pub struct OldRewardPool<B> {
+        pub balance: B,
+        pub total_earnings: B,
+        pub points: U256,
+    }
+
+    enum EitherRewardPool<T: Config, B> {
+        Old(OldRewardPool<B>),
+        New(RewardPool<T>),
+    }
+
+    impl<T: Config, B: Decode> Decode for EitherRewardPool<T, B> {
+        fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+            let len = input.remaining_len()?.unwrap_or_default();
```

Comment:
This is not how `decode` should be implemented. The idea of codec is that you decode by reading the stream until you read enough bytes. The remaining bytes in the stream likely correspond to some other objects. I understand this is only to hack this migration through, but still...

This is only used in `translate`, correct?

What is `translate` using underneath? Is it using `decode_all`?

Reply:
Also you should use `decode_all` everywhere, I guess.

Reply:
This is only used in the translate, changed it to use decode_all

Reply:
Heh. The only reason this works is because the old format uses more bytes than the new one, AND, you are trying in the order `Old`, `New`. If you tried `New`, `Old`, then it wouldn't work :P. I just wanted to make sure we all understand this, and that this is walking on thin ice :)

Reply:
now with decode_all in the decode impl order should not matter, right?

Reply:
Correct.